### PR TITLE
bpo-40566: Apply PEP 573 to abc module

### PR DESCRIPTION
**Pull request by corona10**
https://bugs.python.org/issue40566

Automerge-Triggered-By: @tiran

### CODE REVIEW

File: Modules/_abc.c [hunk 1]
```diff
@@ -629,15 +623,16 @@ _abc__abc_subclasscheck_impl(PyObject *module, PyObject *self,
         goto end;
     }
 
+    _abcmodule_state *state = get_abc_state(module);
```
Comment (tiran) [inline Modules/_abc.c]: Some compilers don't like variable definitions in the middle of a code block. Please define the variable at the top but keep the assignment here.

File: Modules/_abc.c [hunk 2]
```diff
@@ -88,7 +81,8 @@ abc_data_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
     self->_abc_registry = NULL;
     self->_abc_cache = NULL;
     self->_abc_negative_cache = NULL;
-    self->_abc_negative_cache_version = abc_invalidation_counter;
+    _abcmodule_state * state = PyType_GetModuleState(type);
```
Comment (tiran) [inline Modules/_abc.c]: Some compilers don't like variable definitions in the middle of a code block. Please define the variable at the top but keep the assignment here.

You need to check the result for ``NULL`` and handle the error case.

**Review by tiran — CHANGES_REQUESTED** [review]

**Review by corona10 — COMMENTED** [review]
@tiran 
Thanks for the review :)
I have made the requested changes; please review again

File: Modules/_abc.c [hunk 3]
```diff
@@ -81,14 +74,15 @@ static PyObject *
 abc_data_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
 {
     _abc_data *self = (_abc_data *) type->tp_alloc(type, 0);
-    if (self == NULL) {
+    _abcmodule_state *state = PyType_GetModuleState(type);
+    if (self == NULL || state == NULL) {
         return NULL;
```
Comment (tiran) [inline Modules/_abc.c:85]: This may leak memory in case ``tp_alloc()`` is succesful but ``PyType_GetModuleState`` fails.

**Review by tiran — CHANGES_REQUESTED** [review]

File: Modules/_abc.c (hunk 3 above)
Comment (corona10) [reply Modules/_abc.c:85]: Nice catch Thank you :)

**Review by corona10 — COMMENTED** [review]
I have made the requested changes; please review again

**Comment by corona10** [comment]
@tiran

**Review by tiran — APPROVED** [review]

**Review by tiran — APPROVED** [review]
Thanks!
Congratulations, you have survived the sharp edges of C :1st_place_medal:.

**Comment by vstinner** [comment]
Oh, I'm not sure if comments on merged commits are seen by authors. So I copy/paste it here.

state cannot be NULL. If the memory allocation to allocate the module state fails, the module initialization fails (return NULL). If PyType_FromModuleAndSpec() fails to create the type, it returns NULL.

state can only be NULL if the type was not created by PyType_FromModuleAndSpec() or if the module has no state. It's not the case here.

**Comment by corona10** [comment]
@vstinner 
Thanks for the comment.
I will submit the PR about that. :)

**Comment by tiran** [comment]
But ``PyType_GetModuleState()`` can return NULL and therefore the caller should also check for NULL.

**Comment by vstinner** [comment]
> At least use assert() to teach code analysers that the return value cannot be NULL.

Adding "type var = ...; assert(var ...);" would introduce a compiler warning "unused variable 'var' ..." which is annoying :-(

> At least use assert() to teach code analysers that the return value cannot be NULL.

I don't think that we should spend too much time trying to make all code analyzers happy when they report false alarms. It's usually trivial to make a false alarm quiet in the code analyzer.

I don't think that we should add a test which eats CPU time at runtime just to make a code analyzer happy. An assertion doesn't have such issue, but it adds a compiler warning...

--

My concern is that calling Py_DECREF(self); in a constructor before initializing all self fields can lead to a crash. If you really want to keep the test, please reorganize the code to only call Py_DECREF(self) after all fields are set.

**Comment by vstinner** [comment]
Oh, I try to get the state *before* creating self, which avoids the Py_DECREF issue.

**Pull request merged**